toolchain/base/kind_switch.h

```diff
@@ -50,65 +50,116 @@ template <typename... Ts>
 template <typename T>
 concept IsStdVariant = IsStdVariantValue<std::decay_t<T>>;
 
-template <typename... Ts>
-struct TypePack {};
-
-template <typename T>
-struct StdVariantTypePackValue;
+#define CARBON_INTERNAL_IDENTIFIER(name) T##name
+// Turns a list of numbers into a list `T0, T1, ...`.
+#define CARBON_INTERNAL_IDENTIFIERS(...)                             \
+  CARBON_FOR_EACH(CARBON_INTERNAL_IDENTIFIER, CARBON_FOR_EACH_COMMA, \
+                  __VA_ARGS__)
+
+#define CARBON_INTERNAL_TYPENAME(name) typename CARBON_INTERNAL_IDENTIFIER(name)
+// Turns a list of numbers into a list `typename T0, typename T1, ...`.
+#define CARBON_INTERNAL_TYPENAMES(...) \
+  CARBON_FOR_EACH(CARBON_INTERNAL_TYPENAME, CARBON_FOR_EACH_COMMA, __VA_ARGS__)
+
+#define CARBON_INTERNAL_ENUM_NAME(n) VariantType##n##NotHandledInSwitch
+// Turns a list of numbers into a list `VariantType0NotHandledInSwitch, ...`.
+#define CARBON_INTERNAL_TYPES_TO_ENUM_NAMES(...) \
+  CARBON_FOR_EACH(CARBON_INTERNAL_ENUM_NAME, CARBON_FOR_EACH_COMMA, __VA_ARGS__)
+
+// Turns a list of numbers into a set to template specializations of the
+// variable `EnumType EnumValue`, with each specialization having the Nth value
+// in the EnumType (as defined by CARBON_INTERNAL_TYPES_TO_ENUM_NAMES).
+#define CARBON_INTERNAL_TYPE_TO_ENUM_NAME(n)                    \
+  template <>                                                   \
+  constexpr EnumType EnumValue<CARBON_INTERNAL_IDENTIFIER(n)> = \
+      EnumType::CARBON_INTERNAL_ENUM_NAME(n)
 
-template <typename... Ts>
-struct StdVariantTypePackValue<std::variant<Ts...>> {
-  using Types = TypePack<Ts...>;
-};
+template <class T>
+concept TypeFoundInVariant = false;
```

Comment:
This is a bit surprising, given it doesn't do what its name implies it would. I think a comment explaining that this is used only to improve the diagnostics would help.

Reply:
Yeah, it was inverted before but that made the diagnostic confusing instead :) Commenting.